File: codes/leetcode/DesignHashset.cpp
class MyHashSet
{
public:
    /** Initialize your data structure here. */
    int totalBuckets;
    vector<vector<int>> buckets;

    MyHashSet()
    {
        totalBuckets = 223;
        buckets = vector<vector<int>>(totalBuckets, vector<int>{});
    }

    int hash(int key)
    {
        return key % totalBuckets;
    }

    void add(int key)
    {
        int index = hash(key);
        if (find(buckets[index].begin(), buckets[index].end(), key) == buckets[index].end())
            buckets[index].push_back(key);
    }

    void remove(int key)
    {
        int index = hash(key);
        if (find(buckets[index].begin(), buckets[index].end(), key) != buckets[index].end())
            buckets[index].erase(find(buckets[index].begin(), buckets[index].end(), key));
    }

    bool contains(int key)
    {
        int index = hash(key);
        if (find(buckets[index].begin(), buckets[index].end(), key) == buckets[index].end())
            return false;
        return true;
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: codes/leetcode/DesignHashSet.py
# ************************************************
# username    :   smmehrab
# fullname    :   s.m.mehrabul islam
# email       :   mehrab.24csedu.001@gmail.com
# institute   :   university of dhaka, bangladesh
# session     :   2017-2018
# ************************************************

class Node:
    def __init__(self, val, next=None) -> None:
        self.val = val
        self.next = next

class MyHashSet:

    def __init__(self):
        self._size = 1000000
        self._set = [Node(0) for _ in range(10000)]

    def add(self, key: int) -> None:
        index = key % self._size
        node = self._set[index]
        while node.next:
            if node.next.val == key:
                return
            node = node.next
        node.next = Node(key)

    def remove(self, key: int) -> None:
        index = key % self._size
        node = self._set[index]
        while node.next:
            if node.next.val == key:
                node.next = node.next.next
                return
            node = node.next

    def contains(self, key: int) -> bool:
        index = key % self._size
        node = self._set[index]
        while node.next:
            if node.next.val == key:
                node.next = node.next.next
                return True
            node = node.next
        return False

# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)